common/nymcoconut/src/proofs/mod.rs

```diff
@@ -0,0 +1,680 @@
+                .iter()
+                .map(|ciphertext| ciphertext.c2()),
+            self.response_keys.iter(),
+            self.response_attributes.iter()
+        )
+        .map(|(c2, res_key, res_attr)| c2 * self.challenge + pub_key * res_key + h * res_attr)
+        .map(|witness| witness.to_bytes())
+        .collect::<Vec<_>>();
+
+        // Cw = (cm * c) + (rr * g1) + (rm[0] * hs[0]) + ... + (rm[n] * hs[n])
+        let commitment_attributes = commitment * self.challenge
+            + g1 * self.response_opening
+            + self
+                .response_attributes
+                .iter()
+                .zip(params.gen_hs().iter())
+                .map(|(res_attr, hs)| hs * res_attr)
+                .sum::<G1Projective>();
+
+        let ciphertexts_bytes = attributes_ciphertexts
+            .iter()
+            .map(|c| c.to_bytes())
+            .collect::<Vec<_>>();
+
+        // re-compute the challenge
+        let challenge = compute_challenge::<ChallengeDigest, _, _>(
+            std::iter::once(params.gen1().to_bytes().as_ref())
+                .chain(hs_bytes.iter().map(|hs| hs.as_ref()))
+                .chain(std::iter::once(h.to_bytes().as_ref()))
+                .chain(std::iter::once(pub_key.to_bytes().as_ref()))
+                .chain(std::iter::once(commitment.to_bytes().as_ref()))
+                .chain(std::iter::once(commitment_attributes.to_bytes().as_ref()))
+                .chain(std::iter::once(
+                    commitment_private_key_elgamal.to_bytes().as_ref(),
+                ))
+                .chain(commitment_keys1_bytes.iter().map(|aw| aw.as_ref()))
+                .chain(commitment_keys2_bytes.iter().map(|bw| bw.as_ref()))
+                .chain(ciphertexts_bytes.iter().map(|c| c.as_ref())),
+        );
+
+        challenge == self.challenge
+    }
+
+    // challenge || rr || rk.len() || rk || rm.len() || rm
+    pub(crate) fn to_bytes(&self) -> Vec<u8> {
+        let keys_len = self.response_keys.len() as u64;
+        let attributes_len = self.response_attributes.len() as u64;
+
+        let mut bytes = Vec::with_capacity(16 + (keys_len + attributes_len + 3) as usize * 32);
+
+        bytes.extend_from_slice(&self.challenge.to_bytes());
+        bytes.extend_from_slice(&self.response_opening.to_bytes());
+        bytes.extend_from_slice(&self.response_private_elgamal_key.to_bytes());
+        bytes.extend_from_slice(&keys_len.to_le_bytes());
+
+        for rk in &self.response_keys {
+            bytes.extend_from_slice(&rk.to_bytes());
+        }
+
+        bytes.extend_from_slice(&attributes_len.to_le_bytes());
+
+        for rm in &self.response_attributes {
+            bytes.extend_from_slice(&rm.to_bytes());
+        }
+
+        bytes
+    }
+
+    pub(crate) fn from_bytes(bytes: &[u8]) -> Result<Self> {
+        // at the very minimum there must be a single attribute being proven
+        if bytes.len() < 32 * 4 + 16 || (bytes.len() - 16) % 32 != 0 {
```

Comment:
```suggestion
        if bytes.len() < 32 * 5 + 16 || (bytes.len() - 16) % 32 != 0 {
```
Since we require one additional response